cpp/src/arrow/dataset/partition.cc

```diff
@@ -780,7 +780,7 @@ Result<PartitionPathFormat> HivePartitioning::FormatValues(
       // field_index <-> path nesting relation
       segments[i] = name + "=" + hive_options_.null_fallback;
     } else {
-      segments[i] = name + "=" + values[i]->ToString();
+      segments[i] = name + "=" + arrow::internal::UriEscape(values[i]->ToString());
```

Comment:
Can you add tests on the C++ side to make sure it is covered?

Reply:
Sure I will add